### ISSUE TITLE
Support extension of existing needs

### ISSUE THREAD

**Issue by twodrops** [useblocks/sphinxcontrib-needs#282]
Support the possibility to add/overwrite options of existing needs.

For example,

```
.. comp-sys::
   :id: COMP_MY_COMPONENT
```

```
.. needs-extend:: 
   :filter: id==COMP_MY_COMPONENT
   :derives: REQ_SYS_1
```

**Comment by Thomas-Swindells-t** [comment]
This is something I've been pondering.
I think there are probably two similar but distinct use cases.

1. you are writing documentation and want to extend the definition of a particular single need. 

Perhaps the following syntax would work

> .. +comp-sys::
>    :id: COMP_MY_COMPONENT
>    :merge-rank: 3
>  
>    another paragraph of content about this component

This could have semantics that at most one instance can be declared without the + (and becomes the main need). Other +needs would be merged into the base, sorted by merge-rank (default 1 if not specified). For rank ties I'd suggest a sort order of declaring document path (so shorter paths naturally come earlier.

2. You want to augment potentially many needs. In this case potentially the filter function could have direct access to read and modify the needs (presumably after they have been merged together).

This would also have the potential to be allowed to generate new needs, for instance perhaps you have a relationship a->b->c->d and you want to make a new relationship a->X->d which aggregates all the combinations of b->c; or count needs matching a certain criteria and then make a statistic need representing this.

**Comment by danwos** [comment]
Great summary about the different use cases.

I think use case 2 covers somehow use case 1, as a filter could define a single need, e.g. by ```id == SPEC_123```.
There may be a short cut for this by a new option, e.g. ```:extend_id: SPEC_123``` or short ```:eid: SPEC_123```.

Merging needs into a new one sounds interesting, but I can't imagine a real life use case, which wouldn't make writing docs too complex.
And overall the merged needs would still exist in the docs, so it somehow doubles the information.
But I'm totally open for examples, how this could help / speed up things.

I'm not a big fan of the ```+comp-sys``` syntax, as this kind of information is not needed, if the target-need is referenced by an id.
But I see some advantages in reading/understanding the sources, as you already get some important information about the target-need (here its type) and not only the filter/id.

I also see a problem, if content shall get extended, replaced or even deleted, e.g. for content.
Maybe one tag shall be added, and one of three already existing tags shall get removed.
So a "patch" syntax in the content part will provide more power (like git patches).
But it will not be so comfortable as overwriting everything by setting an option.

**So please give some feedback, what of the following functions is needed in your cases**:

- Options: 
  - Add new element
  - Overwrite everything
  - Overwrite/Replace specific element
  - Delete single element
- Content: 
  - Add new content (before/after/in the middle)
  - Replace content
  - Remove content
-  Argument(s) (used as title)
   - Extend (start, end, mid)
   - Replace

**Comment by twodrops** [comment]
Wow, this is developing into a really nice discussion.

The merge concept being discussed is very similar to a concept in AUTOSAR called splittable. It is about having elements splitted across multiple files and a merge function constructs the final element by merging the attributes together. It became such a complex topic in the end as there are a lot of corner cases to take care of. I guess we should keep this rather simple and do not address usecases which are not foreseen. 

Perhaps the +syntax we should avoid as it does not seem very natural within sphinx-needs. The merge-rank however I feel is an important point. I would like to avoid specifying it, but what would be the order in which the elements are merged if we do not have it, especially if the needs and the extensions are split across files?

From the list in the end, my current requirement is only "options" and "add new element" to start with.

**Comment by Thomas-Swindells-t** [comment]
Looking further I agree the +syntax isn't very natural sphinx.

I also agree simplicity is important. The scenarios I was imagining were:

- Adding elements to a need: A security review document analysis the component and you want to add a "securityReviewed" flag on component
- Amending list elements on a need: A design mentions the component and you want to add a tag to track that component is related to the feature
- Replacing element (?). A release note mentions a new version of the component has been deployed and the versionDeployed element of the component should be updated. (alternative would be to use seperate flags for each state but I imagine there are times where that wouldn't be ideal)
- Amending content. A new design document adds some details about a component, you want to have it duplicated on the component. (An alternative may be making it a child need of the component instead and then the main component expands these out.

As I mentioned above one tie break criteria could be the full docname (including directory) and line number. This would mean when tied needs in higher level folders are ranked above those in their subfolders which I imagine is what you'd want most of the time (depth first search effectively). Its the type of thing where the algorithm could get more complicated (breath first or by taking hints from instances higher up in the chain) but keeping it simple and easy to explain is probably a good plan.

---------------------------

The use case I've got around creating derived needs is around making it easier to generate summarized diagrams or summarized data.

In the current structure I'm working with I have usecase needs that contain steps. Each step then provides a link between two subcomponents, along with metadata (what the call is).  These is particularly suited for generating sequence diagrams, and I've also been working on a patch for the flow diagram to allow some needs to act as links rather than nodes.

However these diagrams are low level and detailed, some of the time I may want to show which parent components talk to each other and due to which use-case. This would be the tuple (sourceParentComponent, use case ref/title, targetParentComponent).   
Without the usecase ref just having a way to augment the links would be sufficient, but with the metadata a whole new need would be needed to represent it. This could then be included and filtered in tables or other diagram calls as necessary

**Comment by twodrops** [comment]
@Thomas-Swindells-t Thanks for explaining your usecases in detail. Pretty amazing to see the kind of things you are doing with sphinx-needs :)

To get us started I am thinking of an even simpler implementation.

**Support only a single extension of an existing need**

This means you may create a need

```
.. need:
   :id: MY_NEED
```

and extend it exactly once in another file.

```
.. needs-extend:: 
   :filter: id==MY_NEED
   :links: ANOTHER_NEED
```

This could get rid of the ranking problem and having to use a depth-first approach.

We may apply the following rules:
* `needs-extend` can match an existing `need` only once (throw an error for the need if an extension is used multiple times, similar to duplicate `need` being created)
* need options in `needs-extend` has priority over options in `need` (add, update of options is possible with this, for deleting an option an empty option has to be specified in `needs-extend`)
* If `needs-extend` if found first, create it as a `need-proxy` until the `need` is found and created
* If `needs-extend`s (that is, `need-proxy`  internally) exist without a matching `need`, throw a warning

**Comment by danwos** [comment]
Thanks for this really helpful discussion and all the ideas.
They helped a lot to finally come up with a final spec.
As I needed  to make some tests to prove, if some details are possible in sphinx, there is already some really rough documentation (on feature branch only) and a first implementation of the directive:
![image](https://user-images.githubusercontent.com/998700/123287838-56a7df00-d50f-11eb-96da-ef40a0e9fb01.png)

As you can see, the idea from @Thomas-Swindells-t  with the ``+``-syntax got a revival, but on the option level.

So deleting a complete option (``-``) and adding additional input to an option (``+``) are supported.
Modifying single values will not be supported in the initial version. Because I'm also not sure that there is a strong use case for this and that no simple workaround is possible.

Please let me know, what you think.
Will finalize this in the next days for upcoming release ``0.7.0``

**Comment by twodrops** [comment]
Cool. I will play around with it :)
Short question: As discussed, can the needextend be used to extend a need only once?

**Comment by danwos** [comment]
Nothing ready to play around yet.
Only the directive exists and can be configured. But no modification functionality right now.

**Comment by twodrops** [comment]
ah Ok. Thanks for the info. I misinterpreted it when you said that the directive is available on a feature branch.
Let me give some more thoughts and I will give you a feedback on the concept.

Do you plan to support `needextend` once per `need` or what are your thoughts there?

**Comment by twodrops** [comment]
Here are my thoughts:

* I am still wondering if there is a better syntax possible than "+" or "-" for options. For a Language Server Protocol implementation that we have, it might provide a bit tricky to support it, but is doable. I did not find a better approach to do it inorder to support, add, remove and replace (remove+add) of options.
* Shall we have a additional syntax like below as well

```
.. needextend::
  :status: in progress
```

which would be a shortcut for

```
.. needextend::
  :-status:
  :+status: in progress
 ```
If someone wants to do an add (newly) or replace, the shortform would be helpful.

If we use the above syntax, we can start with it, and include the "+" and "-" as a second step (incase these usecases are not urgent). I do not think, it will be a breaking change to support "+" and "-" later.

**Comment by danwos** [comment]
Updated the PR and added a first Prototype of a working ``needextend``.
Docs got also updated.
Only test cases are missing, will add next days.

``+option``, ``-option`` and also ``option`` are supported now.
``needextend`` can be executed on a need several times.
It is in the responsibility of the user to do this wisely :)
To support this, ``is_modified`` and ``modifications`` values are now available on each need, which can be used to filter for modified needs and needs, which maybe got modified too often.

**Comment by danwos** [comment]
I activated rtd for the feature branch.
So the current doc can be found here: https://sphinxcontrib-needs.readthedocs.io/en/needextend/directives/needextend.html

**Comment by Thomas-Swindells-t** [comment]
Its looking good.
Two thoughts (which could be done later)
1. have a shorthand for targeting a single need, perhaps a filter expression just containing a single token that is not a reserved word is evaluated as matching against complete_id (and/or id?)
2. I think link handling could be automated, if the field being updated is in the needs_extra_links then if a new need_id is being added look up that need and populate the _back field, if the need_id is being removed, remove it from the target _back field.

**Comment by danwos** [comment]
Good points.
Implemented them already.
Not sure, if dynamic functions are fully supported, especially for the link handling in needextend.

A single need can be simply extended  by:
```rst
... needextend:: MY_NEED_ID
    :status: new status
```
See [docs](https://sphinxcontrib-needs.readthedocs.io/en/needextend/directives/needextend.html#single-need-modification).

Also link handling is now down in both directions.

Tests are still missing, so PR will need some more time.

**Comment by Thomas-Swindells-t** [comment]
Thank you for all the work on it, its looking really good and I am looking forward to be able to use it.

**Comment by twodrops** [comment]
@danwos  Thanks once again for this cool feature. The basic functionality looks good. I detected one critical bug. I will open an issue.

**Comment by twodrops** [comment]
@danwos If I want to restrict that a `need` object is not modified more than once using `needextend`, would it be the right way to create a `needs_warnings` and checking if the `modifications` count is more than one?
This is to prevent unforeseen behavior in our case, when `needextend` is used from several pages multiple times and the page order determines the final `need` contents (not a problem for ` :+option:` usage but for `:option:` and `:-option:` usage)

**Comment by danwos** [comment]
Right, checking this constraint must be done by the user and I think using ``needs_warnings`` is the right way.

We may think about a better modification monitoring.
E.g. we could count which options got changed how often.
So two different ``needextend`` on the same need would be okay, if they have modified different options.
But I'm not sure, if this is a real or overthought use case...
